pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AbstractAntlrVisitor.java

```diff
@@ -0,0 +1,85 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.antlr;
+
+import java.util.List;
+
+import org.antlr.v4.runtime.tree.ErrorNode;
+import org.antlr.v4.runtime.tree.ParseTree;
+import org.antlr.v4.runtime.tree.ParseTreeVisitor;
+import org.antlr.v4.runtime.tree.RuleNode;
+import org.antlr.v4.runtime.tree.TerminalNode;
+
+import net.sourceforge.pmd.RuleContext;
+import net.sourceforge.pmd.lang.ast.AntlrBaseNode;
+import net.sourceforge.pmd.lang.ast.Node;
+import net.sourceforge.pmd.lang.rule.AbstractRule;
+
+public abstract class AbstractAntlrVisitor<T> extends AbstractRule implements ParseTreeVisitor<T> {
+
+    protected RuleContext data;
+
+    @Override
+    public void start(RuleContext ctx) {
+        data = ctx;
+    }
+
+    @Override
+    public T visit(ParseTree tree) {
+        return tree.accept(this);
+    }
+
+    @Override
+    public T visitChildren(RuleNode node) {
+        T result = this.defaultResult();
+        int n = node.getChildCount();
+
+        for (int i = 0; i < n && this.shouldVisitNextChild(node, result); ++i) {
+            ParseTree c = node.getChild(i);
+            T childResult = c.accept(this);
+            result = this.aggregateResult(result, childResult);
```

Comment:
I agree that we need an abstraction over common visit functionality (maybe composition would be better than inheritance). 

Since this is a base implementation, I'm not sure if we would like to eat up the return value and prevent children from this class to use it. If we are sure we are NEVER going to need it, we can change the return type, but since we don't know all the uses cases this might have, i would leave that choice to the reviewers.

Reply:
I meant to create another visitor class, this one may also be useful indeed. I think though, that that can be handled in another PR and your current implementation is good.